### ISSUE TITLE
Failure to compile on Debian Sid

### ISSUE THREAD

**Issue by everamzah** [minetest/minetest#4446]
```
[ 63%] Building CXX object src/CMakeFiles/minetest.dir/gettext.cpp.o
In file included from /home/james/Games/Minetest/minetest/src/gettext.cpp:24:0:
/usr/include/libintl.h:39:14: error: expected unqualified-id before ‘char’
 extern char *gettext (const char *__msgid)
              ^
/usr/include/libintl.h:39:14: error: expected initializer before ‘char’
src/CMakeFiles/minetest.dir/build.make:3038: recipe for target 'src/CMakeFiles/minetest.dir/gettext.cpp.o' failed
make[2]: *** [src/CMakeFiles/minetest.dir/gettext.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
CMakeFiles/Makefile2:218: recipe for target 'src/CMakeFiles/minetest.dir/all' failed
make[1]: *** [src/CMakeFiles/minetest.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
```

Reverting b11720af459d44a553cb5d23ef776a632fb30a65 results in no build error, however.

**Comment by est31** [comment]
Its a bit weird that this happens, and that it is introduced by b11720a...

@everamzah can you give the output of the compilation if you edit gettext.h and replace in line 28 the `#define gettext(String) String` with `#error "should not be reached"` ?

**Comment by zing269** [comment]
I had this same issue on Arch with gcc 6, no issue using gcc 5.

**Comment by everamzah** [comment]
```
[ 29%] Building CXX object src/CMakeFiles/minetestserver.dir/gettext.cpp.o
In file included from /home/james/Games/Minetest/minetest/src/gettext.cpp:24:0:
/home/james/Games/Minetest/minetest/src/gettext.h:28:3: error: #error "should not be reached"
  #error "should not be reached"
   ^~~~~
[ 30%] Building CXX object src/CMakeFiles/minetestserver.dir/httpfetch.cpp.o
/home/james/Games/Minetest/minetest/src/gettext.h: In function ‘const wchar_t* wgettext(const char*)’:
/home/james/Games/Minetest/minetest/src/gettext.h:44:35: error: ‘gettext’ was not declared in this scope
  return utf8_to_wide_c(gettext(str));
                                   ^
/home/james/Games/Minetest/minetest/src/gettext.h: In function ‘std::__cxx11::string strgettext(const string&)’:
/home/james/Games/Minetest/minetest/src/gettext.h:49:29: error: ‘gettext’ was not declared in this scope
  return gettext(text.c_str());
                             ^
[ 30%] Building CXX object src/CMakeFiles/minetestserver.dir/inventory.cpp.o
src/CMakeFiles/minetestserver.dir/build.make:2510: recipe for target 'src/CMakeFiles/minetestserver.dir/gettext.cpp.o' failed
make[2]: *** [src/CMakeFiles/minetestserver.dir/gettext.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
CMakeFiles/Makefile2:147: recipe for target 'src/CMakeFiles/minetestserver.dir/all' failed
make[1]: *** [src/CMakeFiles/minetestserver.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by est31** [comment]
This error is really weird...

Okay, @everamzah can you now try to

1. Replace the section in gettext.h with:
```c++
#if USE_GETTEXT
	#warning "libintl is included!"
	#include <libintl.h>
#else
	#warning "macro is set!"
	#define gettext(String) String
#endif
```
and paste the (full) compiler output to a pastebin.

2. Replace the `#include <libintl.h>` with `#define gettext(String) String`, and paste the output.

**Comment by kahrl** [comment]
Can reproduce on Gentoo Linux with gcc 4.9.3, using -std=c++11 and ENABLE_GETTEXT=0. Can _not_ reproduce with ENABLE_GETTEXT=1. I haven't checked yet whether skipping -std=c++11 has any effect.

The bug was reported before in #1702, but that issue was closed for some reason (I'm guessing because that issue's title indicated it was a C++11 related problem and didn't mention ENABLE_GETTEXT, so devs tested compiling with C++11 but with ENABLE_GETTEXT=1, and therefore couldn't reproduce the issue).

Using `g++ -E gettext.cpp` (plus the other compiler options that cmake would add), I was able to trace the problem back to the following:
1. `gettext.cpp` includes `gettext.h`
2. `gettext.h` includes `config.h` which defines `USE_GETTEXT=0`
3. The `#if USE_GETTEXT` branch is not executed, so **`gettext.h` does not include `libintl.h`**
4. The `#else` branch is executed and does **`#define gettext(String) String`**
5. Later in `gettext.cpp`, `util/string.h` is included
6. `util/string.h` includes `<iomanip>`
7. gcc's `<iomanip>` includes `<locale>` which includes `<bits/locale>` which includes `<bits/locale_facets_nonio.h>` which includes `<bits/messages_members.h>` **which includes `<libintl.h>`**
8. `libintl.h` declares the `gettext` function as follows:
`extern char *gettext (const char *__msgid) __THROW __attribute_format_arg__ (1);`
9. Due to the `gettext` macro defined in step 4, the above declaration becomes
`extern char *const char *__msgid __THROW __attribute_format_arg__ (1);`
which is of course a syntax error.

TL;DR gettext.h tries to avoid including libintl.h, but libintl.h is included by accident later and this conflicts with the gettext macro defined in gettext.h.

**Comment by everamzah** [comment]
1. http://pastebin.com/xBCJHrLM
1. http://pastebin.com/RLe4HKf1

**Comment by est31** [comment]
Ahh then it makes sense that the iomanip line of the regressing commit caused the bug. The output posted by @everamzah is consistent with @kahrl 's theory.

Now how to fix it... maybe by #undef'ing the gettext macro at the end of the header, and then adding a function "cstrgettext" into the header which is used instead of gettext throughout the remaining code?

**Comment by kahrl** [comment]
That sounds like it should work, but I don't really like forcing every contributor to learn an obscure custom function name instead of the well known `gettext`. Maybe if the function is called `_` which is a common name for it in other projects.

An alternative could be to play with the header inclusion order. My first idea would be to make sure that one of the offending standard headers (such as `<iomanip>`) is included before the `gettext` macro definition. Of course this must be well commented since it's not obvious why `<iomanip>` should be included there.
```
diff --git a/src/gettext.h b/src/gettext.h
index b8d27f7..29e7a45 100644
--- a/src/gettext.h
+++ b/src/gettext.h
@@ -25,6 +25,13 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #if USE_GETTEXT
        #include <libintl.h>
 #else
+       // In certain environments, some standard headers like <iomanip>
+       // include libintl.h. If libintl.h is included after we define our
+       // gettext macro below, this causes a syntax error at the declaration
+       // of the gettext function in libintl.h. Fix this by including
+       // such a header before defining the macro. See issue #4446.
+       #include <iomanip>
+
        #define gettext(String) String
 #endif
```

My second idea would be to edit each file that includes gettext.h and make sure that gettext.h is included last. But this can be problematic if gettext.h is included from other headers, and is difficult to always remember to check when reviewing PRs in the future, so I prefer the first idea.

**Comment by est31** [comment]
What about always including libintl.h? That would be the most minimal solution

**Comment by kahrl** [comment]
That doesn't work if the user explicitly disabled gettext because they don't have it (e.g. when doing a Windows build).

**Comment by est31** [comment]
Your proposed patch is okay with me as well. If you've tested it I'm giving it a :+1:

**Comment by kahrl** [comment]
@est31 Thanks. Yes, I've tested it and it fixes the build error for me. Made a PR: #4452.

@everamzah Could you test if #4452 fixes it for you, too?

**Comment by marufjs** [comment]
In file included from carParking.cpp:3:
/usr/include/libintl.h:39:14: error: expected initializer before ‘.’ token
   39 | extern char *gettext (const char *__msgid)
      |              ^~~~~~~
make: *** [<builtin>: carParking] Error 1
![image](https://user-images.githubusercontent.com/88562298/143210891-364f7215-6071-4061-b73c-6d4a78a6f2b2.png)
In line 39 I can't see any error. How can I solve it?
my included header files are 
```
#include <iostream>
#include <stdio.h>
#include <conio.h>
#include <fstream>
#include <cstring>
#include <iomanip>
#include <ctime>
```

**Comment by rollerozxa** [comment]
did you just ask for help for your own code in a 4 year old issue about minetest compilation issues? anyways the line 39 it refers to is not in carparking.cpp but in the included header file libintl.h. I don't know why it would error like that though.

**Comment by 0-afflatus** [comment]
weirdspam :D
